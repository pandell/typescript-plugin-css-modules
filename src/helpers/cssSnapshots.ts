import { extractICSS, IICSSExports } from 'icss-utils';
import postcss from 'postcss';
import postcssIcssSelectors from 'postcss-icss-selectors';
import ts_module from 'typescript/lib/tsserverlibrary';
import reserved from 'reserved-words';
import { transformClasses } from './classTransforms';
import { Options } from '../options';

const NOT_CAMELCASE_REGEXP = /[\-_]/;
const processor = postcss(postcssIcssSelectors({ mode: 'local' }));

const classNameToProperty = (className: string) => `'${className}': string;`;
const classNameToNamedExport = (className: string) =>
  `export const ${className}: string;`;

const flattenClassNames = (
  previousValue: string[] = [],
  currentValue: string[],
) => previousValue.concat(currentValue);

export const getClasses = (css: string) => {
  try {
    const processedCss = processor.process(css);
    return extractICSS(processedCss.root).icssExports;
  } catch (e) {
    return {};
  }
};

export const createExports = (classes: IICSSExports, options: Options) => {
  const isCamelCase = (className: string) =>
    !NOT_CAMELCASE_REGEXP.test(className);
  const isReservedWord = (className: string) => !reserved.check(className);

  const processedClasses = Object.keys(classes)
    .map(transformClasses(options.camelCase))
    .reduce(flattenClassNames, []);
  const camelCasedKeys = processedClasses
    .filter(isCamelCase)
    .filter(isReservedWord)
    .map(classNameToNamedExport);

  const defaultExport = `\
declare const classes: {
  ${processedClasses.map(classNameToProperty).join('\n  ')}
};
export default classes;
`;

  if (camelCasedKeys.length) {
    return defaultExport + camelCasedKeys.join('\n') + '\n';
  }
  return defaultExport;
};

export const getDtsSnapshot = (
  ts: typeof ts_module,
  scriptSnapshot: ts.IScriptSnapshot,
  options: Options,
) => {
  const css = scriptSnapshot.getText(0, scriptSnapshot.getLength());
  const classes = getClasses(css);
  const dts = createExports(classes, options);
  return ts.ScriptSnapshot.fromString(dts);
};
