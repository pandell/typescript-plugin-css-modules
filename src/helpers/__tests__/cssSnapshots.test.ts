import { readFileSync } from 'fs';
import { IICSSExports } from 'icss-utils';
import { join } from 'path';
import { createExports, getClasses } from '../cssSnapshots';
import { Options } from '../../options';

const testFileNames = ['test.module.css', 'empty.module.css'];

describe('utils / cssSnapshots', () => {
  testFileNames.forEach((fileName) => {
    let classes: IICSSExports;
    const testFile = readFileSync(
      join(__dirname, 'fixtures', fileName),
      'utf8',
    );

    beforeAll(() => {
      classes = getClasses(testFile);
    });

    describe(`with file '${fileName}'`, () => {
      describe('getClasses', () => {
        it('should return an object matching expected CSS', () => {
          expect(classes).toMatchSnapshot();
        });
      });

      describe('createExports', () => {
        it.each<Options>([{}, { camelCase: true }])(
          'should create an exports file (with options %p)',
          (options) => {
            const exports = createExports(classes, options);
            expect(exports).toMatchSnapshot();
          },
        );
      });
    });
  });
});
