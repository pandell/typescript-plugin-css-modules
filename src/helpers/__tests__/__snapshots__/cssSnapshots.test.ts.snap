// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils / cssSnapshots with file 'empty.module.css' createExports should create an exports file (with options {"camelCase": true}) 1`] = `
"declare const classes: {
  
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.css' createExports should create an exports file (with options {}) 1`] = `
"declare const classes: {
  
};
export default classes;
"
`;

exports[`utils / cssSnapshots with file 'empty.module.css' getClasses should return an object matching expected CSS 1`] = `Object {}`;

exports[`utils / cssSnapshots with file 'test.module.css' createExports should create an exports file (with options {"camelCase": true}) 1`] = `
"declare const classes: {
  'classA': string;
  'ClassB': string;
  'classB': string;
  'class-c': string;
  'classC': string;
  'class_d': string;
  'classD': string;
  'parent': string;
  'childA': string;
  'childB': string;
  'nestedChild': string;
};
export default classes;
export const classA: string;
export const ClassB: string;
export const classB: string;
export const classC: string;
export const classD: string;
export const parent: string;
export const childA: string;
export const childB: string;
export const nestedChild: string;
"
`;

exports[`utils / cssSnapshots with file 'test.module.css' createExports should create an exports file (with options {}) 1`] = `
"declare const classes: {
  'classA': string;
  'ClassB': string;
  'class-c': string;
  'class_d': string;
  'parent': string;
  'childA': string;
  'childB': string;
  'nestedChild': string;
};
export default classes;
export const classA: string;
export const ClassB: string;
export const parent: string;
export const childA: string;
export const childB: string;
export const nestedChild: string;
"
`;

exports[`utils / cssSnapshots with file 'test.module.css' getClasses should return an object matching expected CSS 1`] = `
Object {
  "ClassB": "file__ClassB---2bPVi",
  "childA": "file__childA---1hjQD",
  "childB": "file__childB---pq4Ks",
  "class-c": "file__class-c---DZ1TD",
  "classA": "file__classA---2xcnJ",
  "class_d": "file__class_d---1mwNi",
  "nestedChild": "file__nestedChild---2d15b",
  "parent": "file__parent---1ATMj",
}
`;
